src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,314 @@ using namespace opossum::expression_functional;  // NOLINT
+  // recurse into this node's inputs.
+  for (const auto& input : {node->left_input(), node->right_input()}) {
+    if (!input) continue;
+
+    // Make sure the entry exists, then insert all expressions that the current node needs
+    required_expressions_by_node[input];
+    for (const auto& required_expression : required_expressions) {
+      // Add the columns needed here (and above) if they come from the input node
+      if (input->find_column_id(*required_expression)) {
+        required_expressions_by_node[input].emplace(required_expression);
       }
-      const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
-      pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
     }
 
-    // We cannot create a Table without columns - since Chunks rely on their first column to determine their row count
-    if (pruned_column_ids.size() == node->column_expressions().size() && !pruned_column_ids.empty()) {
-      pruned_column_ids.pop_back();
+    recursive_gather_required_expressions(input, required_expressions_by_node, outputs_visited_by_node);
+  }
+}
+
+void prune_join_node(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node) {
+  // Sometimes, joins are not actually used to combine data but only to check the existence of a tuple in a second
+  // table. Example: SELECT c_name FROM customer, nation WHERE c_nationkey = n_nationkey AND n_name = 'GERMANY'
+  // These joins will be rewritten to semi joins. However, we can only do this if the join is on a unique/primary
+  // key column as non-unique joins could possibly emit a matching line more than once.
+
+  auto join_node = std::dynamic_pointer_cast<JoinNode>(node);
+  if (join_node->join_mode != JoinMode::Inner) return;
+
+  // Check whether the left/right inputs are actually needed by following operators
+  auto left_input_is_used = false;
+  auto right_input_is_used = false;
+  for (const auto& output : node->outputs()) {
+    for (const auto& required_expression : required_expressions_by_node[output]) {
+      if (expression_evaluable_on_lqp(required_expression, *node->left_input())) left_input_is_used = true;
+      if (expression_evaluable_on_lqp(required_expression, *node->right_input())) right_input_is_used = true;
     }
+  }
+  DebugAssert(left_input_is_used || right_input_is_used, "Did not expect a useless join");
+  if (left_input_is_used && right_input_is_used) return;
+
+  // Check whether the join predicates operate on unique columns.
+  auto left_joins_on_unique_column = false;
+  auto right_joins_on_unique_column = false;
+
+  const auto& join_predicates = join_node->join_predicates();
+  for (const auto& join_predicate : join_predicates) {
+    const auto is_unique_column = [](const auto& expression) {
```

Comment:
Why define the lambda in the loop?

Reply:
I like to define it as close to its use as possible both for readability and to limit its scope. Since that is entirely a compile time thing, it does not have any additional cost.